graviti/utility/file.py

```diff
@@ -25,16 +26,50 @@ class File(RemoteFileMixin):
     __slots__ = ("url", "_checksum")
 
     def __init__(self, url: str, checksum: str) -> None:
-        super().__init__("")
-        self.url = READ_ONLY_URL(url)
+        self.url = url
         self._checksum = checksum
 
     def _repr_head(self) -> str:
         return f'{self.__class__.__name__}("{shorten(self._checksum)}")'
 
+    def _urlopen(self) -> UserResponse:
+        if not self.url:
```

Comment:
In what condition `self.url` does not exist?

Reply:
if `self.url` is a null string, this line will raise the error.

Reply:
Is that a normal case when url is an empty string?
If not, we can just remove the check.

Reply:
this check works for illegal null string url. in terms of robustness, I prefer remaining this check

Reply:
If we add the check here, which means we need to add the check for all the string in the code.
Also empty string is an illegal case, wrong url, wrong type are also illegal, that means we need check them all.

If in normal case it will not be empty, there is no need to check it.

Reply:
assuming we remain this check here, the reported error could let us know that the url is null; if the url is not null but illegal, it will cause error when we are trying to open the file. I think this is clearer to distinguish errors.

Reply:
Empty string will also trigger exception when open the file.

Reply:
that's true. but is the exception info caused by null string the same as that of other wrong url strings?

Reply:
You can try it.